Added avatar API and frontend

aliceriot opened the pull request:
#### What are the relevant tickets?

closes #179 

#### What's this PR do?

This allows the user to upload a new avatar. There is a little 'photo' icon on the profile picture, and you can click that to open a dialog that lets you pick and then crop a new avatar.

#### Where should the reviewer start?

I think read through the new code and the changes to the `ProfileImage` component.

NO TESTS YET!

#### How should this be manually tested?

You should be able to update your picture! Cropping should update it to the cropped area. I think it should be fine with any image format (although I've only tested `.jpg` and `.png`).

#### Any background context you want to provide?

#### Screenshots (if appropriate)

![drop](https://cloud.githubusercontent.com/assets/6207644/18724591/f59a2ff0-800a-11e6-9388-0371bc3d95b4.png)

![crop](https://cloud.githubusercontent.com/assets/6207644/18724595/f886fdb0-800a-11e6-98c7-23ea0c6e07fd.png)

![change](https://cloud.githubusercontent.com/assets/6207644/18724599/fbb2e878-800a-11e6-84a0-6698d070b858.png)

#### What GIF best describes this PR or how it makes you feel?

aliceriot commented on static/js/util/files.js:
```diff
@@ -0,0 +1,8 @@
+
+export const dataURLToFile = (uri, name) => (
+  new File([dataURLToBlob(uri)], name)
+);
```
just FYI this will be deleted (didn't end up needing it)

noisecapella commented:
Kind of surprised how little Python code needed to be added, which is great!

aliceriot commented on static/js/actions/image_upload.js:
```diff
@@ -0,0 +1,35 @@
+        dispatch(receivePatchUserPhotoSuccess(username));
+      },
+      error => {
+        dispatch(receivePatchUserPhotoFailure(username, error));
```
oh it can, but you have to pass in a custom callback. forgot to update that here!

aliceriot replied:
actually, do you think it's necessary to do anything with the username here? User's aren't going to be able to update profile images for other users.

noisecapella replied:
You're right, I'd be fine removing it

noisecapella commented on static/js/util/api.js:
```diff
@@ -33,6 +33,29 @@ export function csrfSafeMethod(method: string): boolean {
   return /^(GET|HEAD|OPTIONS|TRACE)$/.test(method);
 }
 
+export function fetchWithCSRF(path: string, init: Object = {}): Promise<*> {
```
We should have `fetchJSONWithCSRF` call this code if that's feasible to reduce code duplication

aliceriot commented:
@noisecapella are you using the same image to test? I noticed that earlier and I think it relates to the browser caching images based on their url. we could append the datetime to the end of the filename before saving to get around that, I think. if I use a bunch of different images I don't see the issue.

noisecapella commented:
Yes I was using the same image actually

aliceriot commented:
@noisecapella I think I addressed all issues. I'm going to start working on tests now, but if you want to take a look and leave some feedback about what I've changed feel free!

noisecapella commented:
I'll wait until the tests are in place, I think the PR is generally in a good place for the most part so I don't have much more to say there

aliceriot commented:
cool - I'll ping you when tests are ready!

aliceriot commented on static/js/util/api.js:
```diff
@@ -33,6 +34,50 @@ export function csrfSafeMethod(method: string): boolean {
+
+const setWith = R.curry((path, valFunc, obj) => (
+  R.set(path, valFunc(), obj)
+));
```
`getCookie` needs to be called at runtime. Just using `R.set` means that `getCookie` gets called whenever the module is loaded, which is before we can set `document.cookie` in our testing code.

aliceriot commented:
I need to 

- [ ] fix failing JS tests
- [ ] test new API function
- [ ] test image upload reducer
- [ ] test the new components

aliceriot commented:
@noisecapella I've added unit tests for the new API functions and the new image upload reducer. How important to you think that tests for the component itself are? Just wondering about what's needed to get this merged.

noisecapella commented:
At the moment, not that important. I'll look at it in the manual testing

aliceriot commented:
alright, in that case this is ready for review

noisecapella commented on profiles/views_test.py:
```diff
@@ -282,6 +294,7 @@ def test_patch_own_profile(self):
             uid="{}_edx".format(new_profile.user.username)
         )
         patch_data = ProfileSerializer().to_representation(new_profile)
+        patch_data.__delitem__('image')
```
Can you change this to `del patch_data['image']`?

aliceriot replied:
ah TIL!

noisecapella commented on static/js/components/ProfileImageUploader.js:
```diff
@@ -0,0 +1,99 @@
+  </Dropzone>
+);
+
+class CropperWrapper extends React.Component {
```
Can we move this into its own file?

aliceriot replied:
sure, I was hoping to avoid using any class declarations but for this I needed a `ref` to grab the cropped data from.

noisecapella commented on static/js/components/ProfileImageUploader.js:
```diff
@@ -0,0 +1,99 @@
+  startPhotoEdit:       (p: File) => void;
+  clearPhotoEdit:       () => void;
+  imageUpload:          ImageUploadState;
+  updateUserPhoto:      (i: string) => Promise<*>;
```
Can we be more specific with the type of `Promise` or is this difficult in Flow?

noisecapella commented on static/js/components/search/LearnerResult_test.js:
```diff
@@ -2,20 +2,41 @@
+    )
   );
 
+  beforeEach(() => {
```
What are the changes in this file for?

aliceriot replied:
the changes here are necessary because `<LearnerResult>` includes a `<ProfileImage>`, which is now a container and not simply a presentational component. So I had to add stuff to this file to avoid errors because there wasn't any store for `<ProfileImage>` to hook into.

noisecapella commented on static/js/store/configureStore.js:
```diff
@@ -69,7 +69,7 @@ export default function configureStore(initialState: ?Object) {
 }
 
-export const configureMainTestStore = (reducer: Reducer<*>) => {
+export const configureMainTestStore = (reducer: Reducer<*,*>) => {
```
What do the two types mean here?

aliceriot replied:
the `Reducer` type takes two type parameters, the `Store` and the `Action`.

aliceriot replied:
As to why it wasn't complaining about this before, I'm not sure.

noisecapella commented:
Code looks good, checking functionality one more time...

noisecapella commented:
Looks great, and I don't see the refresh issue anymore :+1: 

![screenshot from 2016-09-23 14-54-27](https://cloud.githubusercontent.com/assets/863262/18797796/ae59ae60-819d-11e6-90cb-84aabced072a.png)
